File: app/src/main/java/com/rachelleignacio/listbucket/presentation/listeners/ListTouchListenerCallback.java
package com.rachelleignacio.listbucket.presentation.listeners;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

/**
 * Created by rachelleignacio on 3/4/17.
 */

public class ListTouchListenerCallback extends ItemTouchHelper.Callback {
    private final ListAdapterTouchListener listTouchListener;

    public ListTouchListenerCallback(ListAdapterTouchListener listener) {
        listTouchListener = listener;
    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
        return makeMovementFlags(0, swipeFlags);
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        if (viewHolder.getItemViewType() != target.getItemViewType()) {
            return false;
        }
        return true;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return true;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        listTouchListener.onRowDismiss(viewHolder.getAdapterPosition());
    }
}

File: app/src/main/kotlin/com/rachelleignacio/listbucket/presentation/listeners/ListTouchListenerCallback.kt
package com.rachelleignacio.listbucket.presentation.listeners

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper

/**
 * Created by rachelleignacio on 8/23/17.
 */
class ListTouchListenerCallback(private val listTouchListener: ListAdapterTouchListener) : ItemTouchHelper.Callback() {

    override fun isLongPressDragEnabled() = false

    override fun isItemViewSwipeEnabled() = true

    override fun getMovementFlags(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?) =
        makeMovementFlags(0, ItemTouchHelper.START or ItemTouchHelper.END)

    override fun onMove(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder?, target: RecyclerView.ViewHolder?) =
        (viewHolder!!.itemViewType == target!!.itemViewType)

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder?, direction: Int) {
        listTouchListener.onRowDismiss(viewHolder!!.adapterPosition)
    }
}